### ISSUE TITLE
Inconsistent Spot Dimensions

### ISSUE THREAD

**Issue by aahsan045** [bowang-lab/BLEEP#6]
Hi,

I am using the same dataset and your BLEEP code to reproduce your results. However, after training, I found that the ground truth spot expression has a shape of (4992, 3559) whereas the predicted spot expression has a shape of (4992, 256) as in training phase, we trained both image encoder and spot encoder to the dimension of 256. I am assuming that in your BLEEP inference.ipynb and BLEEP main.py, "data/filtered_expression_matrices/3/harmony_matrix.npy" is the same expression matrix that we have used for the BLEEP training part. In the end, I am getting different embedding dimensions. Please share your data (only harmony_matrix) so that I can see where I am wrong or guide me to reproduce the correct results.[

Here is my COLAB file for your reference. 

https://colab.research.google.com/drive/1hp-1CWZbDCgXMf-LXwbcH8HhZQHJHtH-?usp=sharing](url)

**Comment by Xrioen** [comment]
Hi, the data associated with the experiments were in review but are now public, so I have released a version of it on github along with the preprocessing script and other relevant outputs. Please let me know if you run into any additional issues. 

The ground truth spot expression is a spot x gene matrix, where as the embedding from the image and spot encoder are 256d representations which are used for neighbor matching. 
Also, it's recommended to filter the raw expression matrices you are currently using with the tissue_pos_list.csv (the column after barcodes) for detected spots prior to training the model. Please let me know if there are any other questions!

**Comment by aahsan045** [comment]
Hi, thanks for uploading the data. I have used your uploaded data, however, the results are not reproduced. I used **ResNet 50 with 150 epochs, 0.001 learning rate, 512 batch size, and AdamW optimizer with 0.001 weight decay.** The model architecture and the data are the same that you uploaded. However, I am facing a strong overfitting of the model with mean co-relation across gens is around **0.015**.


The code file with outputs is given in the link below. 

https://livemanchesterac-my.sharepoint.com/:u:/g/personal/muhammadahtazaz_ahsan_postgrad_manchester_ac_uk/EbZphs_Xt2JKp5amPc-YE_4BFNm1X-py30aHvtNkySEatQ?e=vg1Qi9

Have you also faced this kind of overfitting while training? As per my understanding, we are saving the best epoch weights with minimum validation loss, so when the model starts overfitting, the weights are captured for an initial number of epochs (where validation loss is minimum) and then loss starts increasing. 

Can you suggest the remedy and explain a bit why this is happening? It is the same code+data you uploaded, but the results are not reproduced.

Thanks and looking forward to hearing from you.

**Comment by Xrioen** [comment]
Hi, I took a brief look at the code. The problem likely occurred at the hvg filtering step. Since your preprocessing script modified the order and total number of gene features of the input matrices, the resulting hvg_matrix.npy features no longer line up across the 4 slices. I would recommend following the preprocessing script I've uploaded. 

Regarding overfitting, the best models tend to occur between epoch 20-80 from my experience. Let me know if you are able to reproduce. 

Edit: To get the filtered matrices from raw matrices, filter spots based on spot barcodes according to each sample's corresponding tissue_pos_list.csv

**Comment by aahsan045** [comment]
Hi, thanks for your reply. 

The outputs shown in the code file (I've attached above in the comment) are produced using your data (that you uploaded yesterday),  i.e. harmony_matrices and tissue_position_list directly into the CLIPDataset() for training and testing.  So, the data produced above at the start of my code (pre-processing) is not used at all.

**Comment by aahsan045** [comment]
Hi, 

I did another iteration. Now seems like the model is not overfitting and converged. However, the mean co-relation is very low (0.0002856....), even lower than the overfitted model (0.015). You can see in the code that I am using your uploaded data (harmony matrices and tissue position list). Also, it is the 4th time, I ran the same model on the same data and got these converged results. It seems like there is an issue with setting up the seed, or the model is not generalizing well. As per your advice, I set the epochs to 100.

See the code in the link:

https://livemanchesterac-my.sharepoint.com/:u:/g/personal/muhammadahtazaz_ahsan_postgrad_manchester_ac_uk/EVWGmr1Otj5EtWCflCtT_SwBHnVRHKvkSsfHGkeuJkR_1A?e=PpaCTX

What do you say?

**Comment by Xrioen** [comment]
My apologies, I didn't realize in your notebook the data was later switched to the ones uploaded. Looking at the results of the first notebook more closely, you did successfully reproduce the results to a large extent. Your best model was around epoch 30 and cell wise correlation looks good. The inference script outputs the mean correlation across all genes but that's not the reported value in the manuscript. The low correlation mean across all genes is expected due to the ill-posed nature of the problem. We only expect models to predict the expression patterns of a small handful of genes faithfully. 

So instead, during evaluation, we focused on reporting and comparing the performances on marker genes, hegs and hvgs due to their practical relevance. I pushed a modified version of the inference code to calculate these. Here are the results from a model I previously trained using the same github code. They are not so far off from the values reported in table 1 in the manuscript, and likely from that of your trained model. As for the remaining difference in results, it may be due to the environment or the way distributed training is performed on the cluster. Of note, training is done using 4 gpus each with 128 batch size without concatenating the logits together at each step. 
![image](https://github.com/bowang-lab/BLEEP/assets/45607572/f0349af2-6084-496b-86ca-09fde04fa32e)

**Comment by aahsan045** [comment]
Hi, thanks for timely reply. I got the following results. Please have a look and suggest what to do. I have updated the inference code that you provided yesterday. Only for marker genes, my model is giving correlation value of 0.170 which is not too far from 0.25.

<img width="530" alt="Screenshot 2024-03-01 at 12 36 10" src="https://github.com/bowang-lab/BLEEP/assets/64263837/8f46eecb-f2b6-4505-a234-d0b043cf8335">

I am using only one GPU (A100) machine to train the model. Should I change the batch size to 128 and reduce the number of epochs, say 35? Also, **Why the model is behaving randomly? For multiple iterations on the same data and giving different results.**

**Comment by Xrioen** [comment]
Hi, the results you got from the 4th run is strange. The model likely collapsed to a trivial solution. I have never experienced this. It may be due to the lack of distributed training where loss is computed from a effective batch size of 128 and then reduced across 4 instances before backpropagation. This likely offers more stability during training and better performance than one instance of batch size 512 due to the smoothed loss target being dependent on the embeddings.

**Comment by aahsan045** [comment]
> Hi, thanks for timely reply. I got the following results. Please have a look and suggest what to do. I have updated the inference code that you provided yesterday. Only for marker genes, my model is giving correlation value of 0.170 which is not too far from 0.25.
> 
> <img alt="Screenshot 2024-03-01 at 12 36 10" width="530" src="https://private-user-images.githubusercontent.com/64263837/309262023-8f46eecb-f2b6-4505-a234-d0b043cf8335.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk2Njg0MDAsIm5iZiI6MTcwOTY2ODEwMCwicGF0aCI6Ii82NDI2MzgzNy8zMDkyNjIwMjMtOGY0NmVlY2ItZjJiNi00NTA1LWEyMzQtZDBiMDQzY2Y4MzM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA1VDE5NDgyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFiMWFmNWI3NmE5YzQ1YWIxMTY0Nzk1Zjg5MjgxYWZhNTcxOGJhNTI5YjJkMWFlMTAzYTQ0MGNjNmRkZDhkZWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.jxStC2xLV8fL8W9qWSl6-sSuiYaRA1RSc7B7bJE9nXg">
> I am using only one GPU (A100) machine to train the model. Should I change the batch size to 128 and reduce the number of epochs, say 35? Also, **Why the model is behaving randomly? For multiple iterations on the same data and giving different results.**

Hi, thanks for the reply again. What do you say about these results (SHOWN ABOVE) ? Also, why mean co-relation is greater than maximum correlation?

Also, I performed another experiment, in which I took 199 genes (common and highly variable across all samples, i.e., intersection of hvgs), and its results are below.  Again, the mean correlation is higher than the maximum. Can you please explain?

<img width="521" alt="image" src="https://github.com/bowang-lab/BLEEP/assets/64263837/06eb99ed-ebb3-4fcc-a5cb-8b683d53e35e">

**Comment by Xrioen** [comment]
Hi, the mean is computed across cell-wise correlations, whereas the max correlation is calculated from correlation gene-wise. Regarding your results above, I suspect it's also partly due to the difference in distributed training. During my experiments I didn't concatenate the embeddings produced from each individual batch of 128 prior to calculating loss, but rather only reduce the loss itself. Hence, it is functionally different from one batch of 512. This may lead to better results compared to your reproduction.

**Comment by aahsan045** [comment]
Hi,

I would like to as a query which is not related to code, however, I would appreciate your opinion on that. 

Given this snapshot of the results that you reproduced and attached in this issue thread above, We see that mean PCC across cells/spots is very high as compared with mean PCC across genes (which is fairly low). Can you comment on why this is happening? I am facing similar behaviour in my experiments using BLEEP. 

![image](https://github.com/user-attachments/assets/888d56dd-2c03-42d6-babe-2c37c4cdbc4b)


Thanks
Looking forward,
